src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2372,16 +2372,35 @@ protected virtual void MarkReflectionLikeDependencies (MethodBody body)
 				if (methodCalled == null)
 					continue;
 
-				var methodCalledDefinition = methodCalled.Resolve ();
-				if (methodCalledDefinition == null)
-					continue;
+				MarkReflectionLikeDependencyForMethodCall (body, methodCalling, instructions, i, methodCalled);
+			}
+		}
 
-				var methodCalledType = methodCalled.DeclaringType;
+		private void MarkReflectionLikeDependencyForMethodCall (
+			MethodBody body,
+			MethodDefinition methodCalling,
+			Collection<Instruction> instructions,
+			int i,
+			MethodReference methodCalled)
+		{ 
+			var methodCalledDefinition = methodCalled.Resolve ();
+			if (methodCalledDefinition == null)
+				return;
+
+			var methodCalledType = methodCalled.DeclaringType;
+
+			// IMPORTANT:
```

Comment:
I'm not sold on this whole defining methods inside methods feature.  This feels horribly abused here.  A single methods with hundreds of lines and a handful of 100 line methods defined inside it.  And setting flags such as `reportedPatternRecognition` defined 600 lines above in the outer methods. 

The way this is structured I have little faith it will succeed it always reporting Recognized or Unrecognized.  It looks trivial to forget to add a call to `ReportUnrecognizedReflectionCallPattern` or `ReportRecognizedReflectionAccessPattern`

Here is the pattern we use in our reflection marking which does a much better job of guaranteeing something or recorded on every usage.  It's not perfect, but it centralizes the responsibility of ensuring everything is recorded as success or failure.

```
            foreach (var reflectionCall in reflectionCalls)
            {
                using (var reflectionContext = new ReflectionUsageContext(Context, modificationContext.Inspector, reflectionCall, _recordedInstructions))
                {
                    var stackItem = modificationContext.Inspector.StackEvaluator.Evaluate(reflectionCall.Instruction);

                    reflectionContext.SetStackEvaluationResult(stackItem);

                    // If we were unable to evaluate the stack then we won't be able to assist with marking of this reflection usage.
                    if (stackItem == null)
                        continue;

                    reflectionCall.Handler(reflectionContext, stackItem);
                }
            }
```

And then a given handler may look like
```
        void ProcessCreateInstance(ReflectionUsageContext reflectionContext, DetailedStackItem stackItem)
        {
            var reflectionMethod = reflectionContext.ReflectionCall.ReflectionMethod;
            // Activator.CreateInstance<T>()
            if (!reflectionMethod.HasParameters && reflectionMethod is GenericInstanceMethod genericInstanceMethod && genericInstanceMethod.GenericArguments.Count == 1)
            {
                // It must be the no arguments overload
                var typeToMark = genericInstanceMethod.GenericArguments[0]?.Resolve();

                if (typeToMark != null)
                {
                    reflectionContext.TraceRecordAndMark(
                        typeToMark,
                        () =>
                        {
                            MarkType(typeToMark);
                            MarkDefaultConstructor(typeToMark);
                        });
                }
                return;
            }

            var loads = stackItem.Loads.Cast<DetailedStackItem>().ToArray();

            // Activator.CreateInstance(Type)
            if (reflectionMethod.Parameters.Count == 1)
            {
                var typeToMark = DetermineType(reflectionContext.Inspector, loads[0])?.Resolve();
                if (typeToMark != null)
                {
                    reflectionContext.TraceRecordAndMark(
                        typeToMark,
                        () =>
                        {
                            MarkType(typeToMark);
                            MarkDefaultConstructor(typeToMark);
                        });
                }
                return;
            }
        }
```

And our context class looks like
```

        class ReflectionUsageContext : IDisposable
        {
            private readonly UnityLinkContext _context;
            public readonly ReflectionCall ReflectionCall;
            public readonly IBodyInspector Inspector;
            private readonly HashSet<Instruction> _recordedInstructions;

            private bool _oneOrMoreMatchesFound;
            private DetailedStackItem _stackItem;

            public ReflectionUsageContext(UnityLinkContext context, IBodyInspector inspector, ReflectionCall reflectionCall, HashSet<Instruction> recordedInstructions)
            {
                Inspector = inspector;
                _context = context;
                ReflectionCall = reflectionCall;
                _recordedInstructions = recordedInstructions;
            }

            public void SetStackEvaluationResult(DetailedStackItem stackItem)
            {
                _stackItem = stackItem;
            }

            public void TraceRecordAndMark<T>(T memberToMark, Action mark)
                where T : IMetadataTokenProvider
            {
                using (_context.Tracer.PushScoped($"Reflection-{memberToMark}"))
                {
                    _oneOrMoreMatchesFound = true;
                    mark();
                    _recordedInstructions.Add(ReflectionCall.Instruction);
                    _context.ReflectionTracker.RecordDetectedUsage(Inspector.Body.Method, memberToMark, ReflectionCall.Instruction, _stackItem);
                }
            }

            public void Dispose()
            {
                // TODO by Mike : Record as undetectable
                if (!_oneOrMoreMatchesFound)
                {
                    _context.ReflectionTracker.RecordUndetectableUsage(Inspector.Body.Method, ReflectionCall.Instruction, _stackItem);
                }
            }
        }
``` 

The important part is that the handling of every reflection usage does not call any Mark methods directly.  Instead it passes an action to `reflectionContext.TraceRecordAndMark` so that the `ReflectionContext` can know at the end if anything was marked or not.  Then we can happily bail at any point and we still ensure the reflection usage is recorded.

I'm not saying this has to structured like ours but I think assuming every break & return will come with an Unrecognized call is tedious.

Reply:
Thanks for the suggestion. I adapted the code to match the pattern you describe above. There's still lot of potential for improvements (like moving the entire reflection pattern detection to a separate source file), but that's for another change (this one is already too big).